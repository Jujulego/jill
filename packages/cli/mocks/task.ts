import { SpawnTask, SpawnTaskOption, TaskStatus } from '@jujulego/jill-core';

// Methods
export class MockTask extends SpawnTask {
  // Constructor
  constructor(cmd: string, opts?: SpawnTaskOption) {
    super(cmd, [], opts);
  }

  // Methods
  _setStatus(status: TaskStatus): void {
    super._setStatus(status);
  }

  _start(): void {
    return;
  }

  _stop(): void {
    return;
  }
}
