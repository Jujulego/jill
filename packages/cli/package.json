{
  "name": "@jujulego/jill",
  "version": "1.1.19",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/Jujulego/jill",
    "directory": "packages/cli"
  },
  "bin": "./bin/jill.js",
  "main": "./dist/main.js",
  "files": [
    "bin",
    "dist"
  ],
  "scripts": {
    "lint": "eslint {src,tests}/**/*.*",
    "clean": "gulp clean",
    "build": "gulp build",
    "watch": "gulp watch",
    "start": "ts-node src/main.ts",
    "test": "jest"
  },
  "engines": {
    "node": "^14"
  },
  "dependencies": {
    "@jujulego/jill-common": "^1.0.0",
    "@jujulego/jill-core": "^1.1.12",
    "@jujulego/jill-myr": "^0.1.0",
    "@repeaterjs/repeater": "^3.0.4",
    "chalk": "^4.1.1",
    "graphql": "^15.7.2",
    "graphql-request": "^3.6.1",
    "graphql-tag": "^2.12.6",
    "graphql-ws": "^5.5.5",
    "ora": "^5.4.1",
    "slugify": "^1.6.0",
    "winston": "^3.3.3",
    "winston-transport": "^4.4.0",
    "ws": "^8.3.0",
    "yargs": "^17.0.1"
  },
  "devDependencies": {
    "@babel/core": "7.16.5",
    "@babel/plugin-proposal-class-properties": "7.16.5",
    "@babel/preset-env": "7.16.5",
    "@babel/runtime": "7.16.5",
    "@types/gulp": "4.0.9",
    "@types/jest": "27.0.3",
    "@types/node": "14.18.3",
    "@types/ws": "8.2.1",
    "@types/yargs": "17.0.8",
    "@typescript-eslint/eslint-plugin": "5.8.1",
    "@typescript-eslint/parser": "5.8.1",
    "del": "6.0.0",
    "eslint": "7.32.0",
    "eslint-plugin-jest": "25.3.2",
    "eslint-plugin-workspaces": "0.7.0",
    "gulp": "4.0.2",
    "gulp-cli": "2.3.0",
    "jest": "27.4.5",
    "jill-tools": "workspace:*",
    "msw": "0.36.3",
    "ts-jest": "27.1.1",
    "ts-node": "10.4.0",
    "typescript": "4.5.4"
  }
}
