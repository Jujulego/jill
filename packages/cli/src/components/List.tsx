import { Box, Static, Text, useStdout } from 'ink';
import { ReactElement } from 'react';

// Types
export type ListItem<K extends string> = Partial<Record<K, string | number>> & Record<string, unknown>;

export interface ListProps<K extends string> {
	attrs: K[];
	data: ListItem<K>[];
	withoutHeaders?: boolean;
	json?: boolean;
}

// Utils
function capitalize(str: string): string {
	return `${str.at(0)?.toUpperCase()}${str.substring(1)}`;
}

// Component
export const List = <K extends string>(props: ListProps<K>): ReactElement => {
	const { attrs, data, withoutHeaders, json } = props;
	const { stdout } = useStdout();

	// Render
	if (json) {
		const lines = JSON.stringify(data, null, stdout?.isTTY ? 2 : 0).split('\n');

		return (
			<Static items={lines}>
				{ (line, idx) => <Text key={idx}>{ line }</Text> }
			</Static>
		);
	}

	return (
		<Box>
			{attrs.map((attr) => (
				<Box key={attr} flexDirection="column" marginRight={2}>
					{!withoutHeaders && <Text bold>{capitalize(attr)}</Text>}
					{data.map((d, idx) => (
						<Text key={idx} color={d[attr] ? '' : 'grey'}>{d[attr] || 'unset'}</Text>
					))}
				</Box>
			))}
		</Box>
	);
};
