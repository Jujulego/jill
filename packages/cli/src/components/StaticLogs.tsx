import { logger } from '@jujulego/jill-core';
import { FC, useEffect, useState } from 'react';
import { format } from 'winston';
import Transport from 'winston-transport';
import { Static, Text } from 'ink';

// Components
export const StaticLogs: FC = () => {
  // State
  const [logs, setLogs] = useState<any[]>([]);

  // Effects
  useEffect(() => {
    logger.add(new class extends Transport {
      // Constructor
      constructor() {
        super({
          format: format.combine(
            format.errors(),
            format.colorize({ message: true, colors: { debug: 'grey', verbose: 'blue', info: 'white', error: 'red' } }),
          )
        });
      }

      // Methods
      log(log: any, next: () => void): void {
        setLogs((old) => [...old, log]);
        next();
      }
    });
  }, []);

  // Render
  return (
    <Static items={logs}>
      { ({ label, message }, idx) => (
        <Text key={idx}>
          <Text color="grey">jill: { label && `[${label}] ` }</Text>
          { message }
        </Text>
      ) }
    </Static>
  );
};
