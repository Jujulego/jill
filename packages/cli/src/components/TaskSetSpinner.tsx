import { Task, TaskSet } from '@jujulego/jill-core';
import { TaskSpinner } from '@jujulego/jill-common';
import { FC, useEffect, useState } from 'react';

// Types
export interface TaskSetSpinnerProps {
  taskSet: TaskSet;
}

// Component
export const TaskSetSpinner: FC<TaskSetSpinnerProps> = (props) => {
  const { taskSet } = props;

  // State
  const [tasks, setTasks] = useState<Task[]>([]);

  // Effects
  useEffect(() => {
    const handler = (task: Task) => {
      setTasks((old) => [...old, task]);
    };

    taskSet.on('started', handler);

    return () => {
      taskSet.off('started', handler);
    };
  }, [taskSet]);

  // Render
  return (
    <>
      { tasks.map((task, idx) => <TaskSpinner key={idx} task={task} />) }
    </>
  );
};
