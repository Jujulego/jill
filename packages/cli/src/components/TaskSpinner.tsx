import { Task } from '@jujulego/jill-core';
import { Text } from 'ink';
import Spinner from 'ink-spinner';
import logSymbols from 'log-symbols';
import { FC, useEffect, useState } from 'react';

// Types
export interface TaskSpinnerProps {
  task: Task;
}

// Component
export const TaskSpinner: FC<TaskSpinnerProps> = (props) => {
  const { task } = props;

  // State
  const [status, setStatus] = useState(task.status);

  // Effects
  useEffect(() => {
    task.on('status', setStatus);

    return () => {
      task.off('status', setStatus);
    };
  }, [task]);

  // Render
  if (status === 'done') {
    return (
      <Text>
        <Text color="green">{ logSymbols.success }</Text>{' '}
        <Text color="grey">[{ task.context.workspace?.name }]</Text>{' '}
        { task.name } is { status }
      </Text>
    );
  } else if (status === 'failed') {
    return (
      <Text>
        <Text color="red">{ logSymbols.error }</Text>{' '}
        <Text color="grey">[{ task.context.workspace?.name }]</Text>{' '}
        { task.name } is { status }
      </Text>
    );
  }

  return (
    <Text>
      <Spinner type={status === 'running' ? 'line' : 'simpleDotsScrolling'} />{' '}
      <Text color="grey">[{ task.context.workspace?.name }]</Text>{' '}
      { task.name } is { status }
    </Text>
  );
};
