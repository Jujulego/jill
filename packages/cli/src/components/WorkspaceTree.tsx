import { Workspace } from '@jujulego/jill-core';
import { Newline, Text, TextProps } from 'ink';
import { FC, ReactElement, useEffect, useState } from 'react';

// Types
export interface WorkspaceTreeProps {
  workspace: Workspace;
  dev?: boolean;
  level?: ReactElement;
}

// Utils
const style = (dev: boolean): TextProps => ({ color: dev ? 'blue' : '' });

// Component
export const WorkspaceTree: FC<WorkspaceTreeProps> = (props) => {
  const { workspace: wks, dev = false, level = '' } = props;

  // State
  const [deps, setDeps] = useState<[Workspace, boolean][]>([]);

  // Effects
  useEffect(() => void (async () => {
    const deps: [Workspace, boolean][] = [];

    for await (const dep of wks.dependencies()) {
      deps.push([dep, dev]);
    }

    for await (const dep of wks.devDependencies()) {
      deps.push([dep, true]);
    }

    setDeps(deps);
  })(), [wks]);

  // Render
  return (
    <Text>
      <Text {...style(dev)}>{ wks.name }</Text>
      { wks.version && (<Text color="grey">@{ wks.version }</Text>) }
      <Newline />

      { deps.map(([dep, isDev], idx) => (
        <Text key={dep.name}>
          { level }<Text {...style(dev)}>{ idx === deps.length - 1 ? '└' : '├'}─{' '}</Text>
          <WorkspaceTree
            workspace={dep}
            dev={isDev}
            level={<>{ level }<Text {...style(dev)}>{ idx === deps.length - 1 ? ' ' : '│' }{'  '}</Text></>}
          />
        </Text>
      )) }
    </Text>
  );
};
