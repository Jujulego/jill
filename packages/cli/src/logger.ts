import { logger as coreLogger } from '@jujulego/jill-core';
import { format, Logger } from 'winston';
import Transport from 'winston-transport';
import ora from 'ora';
import chalk from 'chalk';

// Constants
const MESSAGE = Symbol.for('message');

// Transport
export class OraTransport extends Transport {
  // Attributes
  readonly _spinner = ora();

  // Methods
  private _keepSpinner<T>(fun: () => T): T {
    // Save state
    let spinning = false;
    let text = '';

    if (this._spinner.isSpinning) {
      spinning = true;
      text = this._spinner.text;
    }

    try {
      return fun();
    } finally {
      // Restore state
      if (!this._spinner.isSpinning && spinning) {
        this._spinner.start(text);
      }
    }
  }

  log(info: any, next: () => void): void {
    // Print message
    const msg = info[MESSAGE] as string;

    this._keepSpinner(() => {
      this._spinner.stop();

      for (const line of msg.split('\n')) {
        process.stderr.write(line + '\n');
      }
    });

    next();
  }

  spin(message: string): void {
    this._spinner.start(message);
  }

  succeed(log: string): void {
    this._spinner.succeed(log);
  }

  fail(log: Error | string): void {
    if (typeof log === 'string') {
      this._spinner.fail(log);
    } else {
      this._spinner.fail(log.stack || log.toString());
    }
  }

  stop(): void {
    this._spinner.stop();
  }
}

// Logger
export class OraLogger {
  // Logger
  constructor(
    private readonly logger: Logger,
    private readonly transport: OraTransport
  ) {}

  // Methods
  // - logger
  debug(message: string): void {
    this.logger.debug({ message });
  }

  verbose(message: string): void {
    this.logger.verbose({ message });
  }

  info(message: string): void {
    this.logger.info({ message });
  }

  warn(message: string): void {
    this.logger.warn({ message });
  }

  error(message: string): void {
    this.logger.error({ message });
  }

  child(options: Record<string, unknown>): Logger {
    return this.logger.child(options);
  }

  // - ora
  spin(msg: string): void {
    this.transport.spin(msg);
  }

  succeed(msg: string): void {
    this.transport.succeed(msg);
  }

  fail(msg: string): void {
    this.transport.fail(msg);
  }

  stop(): void {
    this.transport.stop();
  }

  // Properties
  get level(): string {
    return this.logger.level;
  }

  set level(level: string) {
    this.logger.level = level;
  }
}

// Setup
export const transport = new OraTransport({
  format: format.combine(
    format.colorize({ message: true, colors: { debug: 'grey', verbose: 'blue', info: 'white', error: 'red' } }),
    format.printf(({ label, message }) => message.split('\n').map(line => [label && chalk.grey(`[${label}]`), line].filter(p => p).join(' ')).join('\n')),
  )
});

coreLogger.add(transport);

export const logger = new OraLogger(coreLogger, transport);
