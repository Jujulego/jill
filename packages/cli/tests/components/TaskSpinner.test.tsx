import { Project, Workspace } from '@jujulego/jill-core';
import { render, cleanup } from 'ink-testing-library';

import { TaskSpinner } from '../../src/components/TaskSpinner';

import '../logger';
import { pkg } from '../utils/package';
import { TestTask } from '../utils/task';

// Setup
let tsk: TestTask;
let wks: Workspace;

beforeEach(() => {
  tsk = new TestTask('Test TaskSpinner');
  wks = new Workspace('test', pkg({ name: 'test' }), new Project('/'));
});

afterEach(() => {
  cleanup();
});

// Tests
describe('TaskSpinner', () => {
  it('should follow task\'s status', async () => {
    expect(tsk.listenerCount('status')).toBe(0);

    // First render
    const { lastFrame, rerender, unmount } = render(<TaskSpinner task={tsk} />);
    await new Promise(res => setTimeout(res, 0));

    expect(lastFrame()).toMatchSnapshot();
    expect(tsk.listenerCount('status')).toBe(1);

    // Update state
    tsk._setStatus('done');
    rerender(<TaskSpinner task={tsk} />);

    expect(lastFrame()).toMatchSnapshot();

    // Unmount
    unmount();
    await new Promise(res => setTimeout(res, 0));

    expect(tsk.listenerCount('status')).toBe(0);
  });

  it('should print blocked task', () => {
    // Render
    tsk._setStatus('blocked');
    const { lastFrame } = render(<TaskSpinner task={tsk} />);

    expect(lastFrame()).toMatchSnapshot();
  });

  it('should print ready task', () => {
    // Render
    tsk._setStatus('ready');
    const { lastFrame } = render(<TaskSpinner task={tsk} />);

    expect(lastFrame()).toMatchSnapshot();
  });

  it('should print running task', () => {
    // Render
    tsk._setStatus('running');
    const { lastFrame } = render(<TaskSpinner task={tsk} />);

    expect(lastFrame()).toMatchSnapshot();
  });

  it('should print done task', () => {
    // Render
    tsk._setStatus('done');
    const { lastFrame } = render(<TaskSpinner task={tsk} />);

    expect(lastFrame()).toMatchSnapshot();
  });

  it('should print failed task', () => {
    // Render
    tsk._setStatus('failed');
    const { lastFrame } = render(<TaskSpinner task={tsk} />);

    expect(lastFrame()).toMatchSnapshot();
  });

  it('should print blocked workspace task', () => {
    // Render
    tsk._setStatus('blocked');
    tsk._setContext({ workspace: wks });
    const { lastFrame } = render(<TaskSpinner task={tsk} />);

    expect(lastFrame()).toMatchSnapshot();
  });

  it('should print ready workspace task', () => {
    // Render
    tsk._setStatus('ready');
    tsk._setContext({ workspace: wks });
    const { lastFrame } = render(<TaskSpinner task={tsk} />);

    expect(lastFrame()).toMatchSnapshot();
  });

  it('should print running workspace task', () => {
    // Render
    tsk._setStatus('running');
    tsk._setContext({ workspace: wks });
    const { lastFrame } = render(<TaskSpinner task={tsk} />);

    expect(lastFrame()).toMatchSnapshot();
  });

  it('should print done workspace task', () => {
    // Render
    tsk._setStatus('done');
    tsk._setContext({ workspace: wks });
    const { lastFrame } = render(<TaskSpinner task={tsk} />);

    expect(lastFrame()).toMatchSnapshot();
  });

  it('should print failed workspace task', () => {
    // Render
    tsk._setStatus('failed');
    tsk._setContext({ workspace: wks });
    const { lastFrame } = render(<TaskSpinner task={tsk} />);

    expect(lastFrame()).toMatchSnapshot();
  });
});
