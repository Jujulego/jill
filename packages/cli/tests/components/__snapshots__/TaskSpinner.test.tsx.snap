// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskSpinner should follow task's status 1`] = `".   Test TaskSpinner is ready"`;

exports[`TaskSpinner should follow task's status 2`] = `"[32mâˆš[39m Test TaskSpinner is done"`;

exports[`TaskSpinner should print blocked task 1`] = `".   Test TaskSpinner is blocked"`;

exports[`TaskSpinner should print blocked workspace task 1`] = `".   [90m[test][39m Test TaskSpinner is blocked"`;

exports[`TaskSpinner should print done task 1`] = `"[32mâˆš[39m Test TaskSpinner is done"`;

exports[`TaskSpinner should print done workspace task 1`] = `"[32mâˆš[39m [90m[test][39m Test TaskSpinner is done"`;

exports[`TaskSpinner should print failed task 1`] = `"[31mÃ—[39m Test TaskSpinner is failed"`;

exports[`TaskSpinner should print failed workspace task 1`] = `"[31mÃ—[39m [90m[test][39m Test TaskSpinner is failed"`;

exports[`TaskSpinner should print ready task 1`] = `".   Test TaskSpinner is ready"`;

exports[`TaskSpinner should print ready workspace task 1`] = `".   [90m[test][39m Test TaskSpinner is ready"`;

exports[`TaskSpinner should print running task 1`] = `"- Test TaskSpinner is running"`;

exports[`TaskSpinner should print running workspace task 1`] = `"- [90m[test][39m Test TaskSpinner is running"`;
